src/ray/core_worker/core_worker.cc

```diff
@@ -59,27 +59,47 @@ namespace ray {
 
 std::unique_ptr<CoreWorkerProcess> CoreWorkerProcess::instance_;
 
+absl::Mutex CoreWorkerProcess::instance_mutex_;
```

Comment:
What is this lock for and why wasn't it needed before? I'm concerned that this lock seems to get acquired in so many places. We should try to think of a different way to do this or at least benchmark.

Reply:
I'm also pretty confused by this / why adding the ref counting code necessitates this

Reply:
In Java, GC runs in a standalone thread, and we can't control the exact timing of garbage collection. It's possible that an object is garbage collected after `CoreWorkerProcess` is shut down, or even when `CoreWorkerProcess` is being shut down. So we need to make sure that accessing the `CoreWorker` instance via `CoreWorkerProcess::TryGetWorker(worker_id)` is thread-safe.

Reply:
@raulchen and I discussed this offline. Because the thread-safety of core worker APIs is not guaranteed, a worker can easily get crashed if accessing core worker APIs in another threading while the core worker is shutting down. This issue applies to both Java and Python. For now, we assume that users won't do this on purpose. In the long term, we should add some kind of protection in core worker APIs.

In this PR, we'll add a lock at Java level to prevent ref counting GC and `Ray.shutdown()` running simultaneously. In this way, we won't need the additional lock in `core_worker.cc`, so impacts on Python workers should be minimal.